File: extensions/extension/src/main/java/com/autocat/morphe/smartlauncher/extension/ShizukuArchiveHelper.java
package com.autocat.morphe.smartlauncher.extension;

import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.widget.Toast;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.lang.reflect.Method;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Pure-reflection Shizuku privileged app archiving & unarchiving helper.
 * Zero compile-time or runtime dependencies on external libraries to guarantee 100% APK stability.
 */
@SuppressWarnings("unused")
public class ShizukuArchiveHelper {

    private static final String TAG = "ShizukuArchiveHelper";
    public static final int SHIZUKU_REQ_CODE = 1001;

    private static final ExecutorService EXECUTOR = Executors.newSingleThreadExecutor();
    private static final Handler MAIN_HANDLER = new Handler(Looper.getMainLooper());

    public enum Status {
        ACTIVE,
        PERMISSION_REQUIRED,
        NOT_RUNNING
    }

    public static Status getStatus() {
        if (!isShizukuAlive()) {
            return Status.NOT_RUNNING;
        }
        if (!hasPermission()) {
            return Status.PERMISSION_REQUIRED;
        }
        return Status.ACTIVE;
    }

    public static boolean isShizukuAlive() {
        try {
            Class<?> shizukuClass = Class.forName("rikka.shizuku.Shizuku");
            Method pingMethod = shizukuClass.getMethod("pingBinder");
            Boolean isAlive = (Boolean) pingMethod.invoke(null);
            return isAlive != null && isAlive;
        } catch (Throwable t) {
            return false;
        }
    }

    public static boolean hasPermission() {
        try {
            Class<?> shizukuClass = Class.forName("rikka.shizuku.Shizuku");
            Method checkPermMethod = shizukuClass.getMethod("checkSelfPermission");
            Integer perm = (Integer) checkPermMethod.invoke(null);
            return perm != null && perm == PackageManager.PERMISSION_GRANTED;
        } catch (Throwable t) {
            return false;
        }
    }

    public static boolean isShizukuAvailable() {
        return isShizukuAlive() && hasPermission();
    }

    public static void requestShizukuPermission(int requestCode) {
        try {
            Class<?> shizukuClass = Class.forName("rikka.shizuku.Shizuku");
            Method requestMethod = shizukuClass.getMethod("requestPermission", int.class);
            requestMethod.invoke(null, requestCode);
            Log.i(TAG, "Requested Shizuku permission with code " + requestCode);
        } catch (Throwable t) {
            Log.e(TAG, "Failed to request Shizuku permission", t);
        }
    }

    public static void requestPermissionWithFeedback(Context context) {
        if (!isShizukuAlive()) {
            postToast(context, "Shizuku service is not running. Please start Shizuku first.");
            return;
        }
        if (hasPermission()) {
            postToast(context, "Shizuku permission is already granted!");
            return;
        }
        postToast(context, "Requesting Shizuku permission…");
        requestShizukuPermission(SHIZUKU_REQ_CODE);
    }

    public static boolean archivePackage(final String packageName) {
        if (packageName == null || packageName.isEmpty()) return false;
        boolean ok = execShizukuCommand("pm archive " + packageName);
        if (!ok) {
            ok = execShizukuCommand("cmd package archive " + packageName);
        }
        return ok;
    }

    public static boolean unarchivePackage(final String packageName) {
        if (packageName == null || packageName.isEmpty()) return false;
        boolean ok = execShizukuCommand("pm unarchive " + packageName);
        if (!ok) {
            ok = execShizukuCommand("cmd package unarchive " + packageName);
        }
        return ok;
    }

    private static boolean execShizukuCommand(String cmd) {
        if (!isShizukuAvailable() || cmd == null) {
            return false;
        }
        try {
            Class<?> shizukuClass = Class.forName("rikka.shizuku.Shizuku");
            Method newProcessMethod = shizukuClass.getMethod("newProcess", String[].class, String[].class, String.class);
            Process process = (Process) newProcessMethod.invoke(null, new String[]{"sh", "-c", cmd}, null, null);
            if (process != null) {
                StringBuilder output = new StringBuilder();
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        output.append(line).append("\n");
                    }
                }
                try (BufferedReader errReader = new BufferedReader(new InputStreamReader(process.getErrorStream()))) {
                    String line;
                    while ((line = errReader.readLine()) != null) {
                        output.append(line).append("\n");
                    }
                }
                int exitCode = process.waitFor();
                String outStr = output.toString().trim();
                Log.i(TAG, "Shizuku cmd [" + cmd + "] exit=" + exitCode + ", out=" + outStr);
                return exitCode == 0 || outStr.toLowerCase().contains("success");
            }
        } catch (Throwable t) {
            Log.e(TAG, "Shizuku command execution failed for: " + cmd, t);
        }
        return false;
    }

    public static void postToast(final Context context, final String message) {
        if (context == null || message == null) return;
        MAIN_HANDLER.post(new Runnable() {
            @Override
            public void run() {
                try {
                    Toast.makeText(context.getApplicationContext(), message, Toast.LENGTH_SHORT).show();
                } catch (Throwable ignored) {}
            }
        });
    }
}

File: patches/src/main/kotlin/com/autocat/morphe/smartlauncher/helpers/ShizukuArchiveHelper.kt
package com.autocat.morphe.smartlauncher.helpers

import android.content.Context
import android.content.pm.PackageManager
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.widget.Toast
import rikka.shizuku.Shizuku

/**
 * Injected helper class for archiving apps via Shizuku on Smart Launcher 6.
 * Asynchronous process execution to prevent ANRs on the main UI thread.
 */
object ShizukuArchiveHelper {

    private const val TAG = "SmartLauncherMorphe_ShizukuArchive"

    @JvmStatic
    fun isShizukuAvailable(): Boolean {
        return try {
            Shizuku.pingBinder() && Shizuku.checkSelfPermission() == PackageManager.PERMISSION_GRANTED
        } catch (e: Throwable) {
            Log.w(TAG, "Shizuku availability check failed", e)
            false
        }
    }

    @JvmStatic
    fun requestShizukuPermission(requestCode: Int) {
        try {
            if (Shizuku.pingBinder() && Shizuku.checkSelfPermission() != PackageManager.PERMISSION_GRANTED) {
                Shizuku.requestPermission(requestCode)
            }
        } catch (e: Throwable) {
            Log.e(TAG, "Failed to request Shizuku permission", e)
        }
    }

    /**
     * Asynchronously archives the given package using Shizuku privileges via `pm archive <packageName>`.
     */
    @JvmStatic
    fun archiveAppWithShizuku(targetObj: Any?, packageName: String): Boolean {
        val context = ArchivedAppFilterHelper.resolveContext(targetObj)
        if (context == null) {
            Log.e(TAG, "Unable to resolve Context from targetObj: $targetObj")
            return false
        }

        if (!isShizukuAvailable()) {
            safeShowToast(context, "Shizuku is not running or permission denied", Toast.LENGTH_SHORT)
            return false
        }

        val mainHandler = Handler(Looper.getMainLooper())
        safeShowToast(context, "Archiving $packageName via Shizuku...", Toast.LENGTH_SHORT)

        Thread {
            try {
                Log.i(TAG, "Attempting Shizuku archive for package: $packageName")
                
                val process = Shizuku.newProcess(
                    arrayOf("pm", "archive", packageName),
                    null,
                    null
                )

                val outputText = process.inputStream.bufferedReader().readText()
                val errorText = process.errorStream.bufferedReader().readText()
                val exitCode = process.waitFor()

                mainHandler.post {
                    if (exitCode == 0) {
                        safeShowToast(context, "App $packageName successfully archived", Toast.LENGTH_SHORT)
                    } else {
                        Log.e(TAG, "Shizuku pm archive failed (exit code $exitCode): $errorText")
                        safeShowToast(context, "Failed to archive app: ${errorText.ifEmpty { outputText }}", Toast.LENGTH_LONG)
                    }
                }
            } catch (e: Throwable) {
                Log.e(TAG, "Error executing Shizuku archive command for $packageName", e)
                mainHandler.post {
                    safeShowToast(context, "Shizuku archive error: ${e.localizedMessage}", Toast.LENGTH_SHORT)
                }
            }
        }.start()

        return true
    }

    private fun safeShowToast(context: Context, text: String, duration: Int) {
        try {
            Toast.makeText(context, text, duration).show()
        } catch (e: Throwable) {
            Log.w(TAG, "Toast display failed: $text", e)
        }
    }
}
